File: python/tree/BalancedBinaryTree.py
from typing import Optional

from tree.TreeNode import TreeNode


class BalancedBinaryTree:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        def dfs(node: Optional[TreeNode]) -> int:
            if not node:
                return 0

            return 1 + max(dfs(node.left), dfs(node.right))

        return (abs(dfs(root.left) - dfs(root.right)) < 2
                and BalancedBinaryTree.isBalanced(self, root.left)
                and BalancedBinaryTree.isBalanced(self, root.right))

File: java/src/prep/leetcode/BalancedBinaryTree.java
package prep.leetcode;

import prep.tree.TreeNode;

public class BalancedBinaryTree {
	public boolean isBalanced(TreeNode root) {
		if(root==null)
			return true;
		
		return Math.abs(dfs(root.left)-dfs(root.right))<2 && isBalanced(root.left) && isBalanced(root.right);
        
    }
	
	public int dfs(TreeNode node) {
		if(node == null)
			return 0;
		
		return 1+ Math.max(dfs(node.left), dfs(node.right));
	}
	
	public static void main(String[] args) {
		TreeNode node1 = new TreeNode(1);
		TreeNode node2 = new TreeNode(2);
		TreeNode node3 = new TreeNode(3);
		TreeNode node4 = new TreeNode(4);
		TreeNode node5 = new TreeNode(5);
		
		node1.left = node2;
		node1.right = node3;
		
		node3.left = node4;
		node3.right = node5;
		
		BalancedBinaryTree bbt = new BalancedBinaryTree();
		System.out.println(bbt.isBalanced(node1));
	}

}
